plugins/gitlab/auth.ml

```diff
@@ -0,0 +1,124 @@
+open Lwt.Infix
+
+module Server = Cohttp_lwt_unix.Server
+
+type t = {
+  client_id : string;
+  client_secret: string;
+  scopes : string list;
+  redirect_uri : string;
+} [@@deriving yojson]
+
+let v ?(scopes=["read_user"]) ~client_id ~client_secret ~redirect_uri () =
+  { client_id; client_secret; scopes; redirect_uri }
+
+exception ScopeOfString of string
+
+let scope_of_string scope =
+  match Gitlab.Scope.of_string scope with
+  | Some s -> s
+  | None -> raise @@ ScopeOfString ("Invalid option for scope_of_string :" ^ scope)
+
+let make_login_uri t ~csrf =
+  Gitlab.Token.create_url
+    ~client_id: t.client_id
+    ~redirect_uri:(Uri.of_string t.redirect_uri)
+    ~state:csrf
+    ~scopes: (List.map scope_of_string t.scopes)
+    ()
+
+let get_access_token t code =
+  Gitlab.Token.of_code
+    ~client_id: t.client_id
+    ~code
+    ~client_secret: t.client_secret
+    ~redirect_uri: t.redirect_uri
+    ()
+
+let get_user token =
+  let open Gitlab in
+  let open Monad in
+  let cmd =
+    User.current_user ~token () >|~ fun user ->
+    Ok ("gitlab:" ^ user.Gitlab_t.user_username)
+  in
+  run cmd
+
+let example_config () =
+  v ~client_id:"..." ~client_secret:"..." ~redirect_uri:"..." ()
+  |> to_yojson
+  |> Yojson.Safe.pretty_to_string
+
+let configuration_howto ctx =
+  Current_web.Context.respond_ok ctx Tyxml.Html.[
+      p [ txt "GitLab single-sign-on has not been configured." ];
+      p [
+        txt "Start the service with ";
+        code [txt "--gitlab-oauth path.json"];
+        txt ", where the file contains:";
+      ];
+      pre [ txt (example_config ()) ]
+    ]
+
+let login t : Current_web.Resource.t = object
+  method get_raw site request =
+    Current_web.Context.of_request ~site request >>= fun ctx ->
+    match t with
+    | None -> configuration_howto ctx
+    | Some t ->
+      let uri = Cohttp.Request.uri request in
+      match Uri.get_query_param uri "code", Uri.get_query_param uri "state" with
+      | None, _ -> Server.respond_error ~status:`Bad_request ~body:"Missing code" ()
+      | _, None -> Server.respond_error ~status:`Bad_request ~body:"Missing state" ()
+      | Some code, Some state ->
+        if state <> Current_web.Context.csrf ctx then (
+          Server.respond_error ~status:`Bad_request ~body:"Bad CSRF token" ()
+        ) else (
+          get_access_token t code >>= function
+          | None ->
+            Server.respond_error ~status:`Internal_server_error ~body:"Failed to get token" ()
+          | Some token ->
+            Log.warn (fun f -> f "Getting the current User");
```

Comment:
Doesn't really sound like a warning to me.

Reply:
Very true, I'll remove it